packages/react-reconciler/src/__tests__/ReactSuspenseEffectsSemantics-test.js

```diff
@@ -2479,13 +2698,106 @@ describe('ReactSuspenseEffectsSemantics', () => {
         ReactNoop.render(null);
       });
       assertLog([
+        'Text:Function destroy insertion',
         'Text:Function destroy layout',
         'ClassText:Class componentWillUnmount',
         'Text:Function destroy passive',
       ]);
     });
   });
 
+  describe('insertion effects within a tree that re-suspends in an update', () => {
+    // @gate enableLegacyCache
+    it('should be destroyed in the deleted tree', async () => {
+      function App({children = null}) {
+        Scheduler.log('App render');
+        React.useInsertionEffect(() => {
+          Scheduler.log('App create insertion');
+          return () => {
+            Scheduler.log('App destroy insertion');
+          };
+        }, []);
+        React.useLayoutEffect(() => {
+          Scheduler.log('App create layout');
+          return () => {
+            Scheduler.log('App destroy layout');
+          };
+        }, []);
+        React.useEffect(() => {
+          Scheduler.log('App create passive');
+          return () => {
+            Scheduler.log('App destroy passive');
+          };
+        }, []);
+        return (
+          <>
+            <Suspense fallback={<Text text="Fallback" />}>{children}</Suspense>
+          </>
+        );
+      }
+
+      await act(() => {
+        ReactNoop.render(
+          <App>
+            <Text text="One" />
+          </App>,
+        );
+      });
+      assertLog([
+        'App render',
+        'Text:One render',
+        'Text:One create insertion',
+        'App create insertion',
+        'Text:One create layout',
+        'App create layout',
+        'Text:One create passive',
+        'App create passive',
+      ]);
+      expect(ReactNoop).toMatchRenderedOutput(<span prop="One" />);
+
+      // Schedule an update that causes React to suspend.
+      await act(async () => {
+        ReactNoop.render(
+          <App>
+            <AsyncText text="Two" ms={1000} />
+          </App>,
+        );
+        await waitFor([
+          'App render',
+          'Suspend:Two',
+          'Text:Fallback render',
+          'Text:One destroy layout',
+          'Text:Fallback create insertion',
+          'Text:Fallback create layout',
+        ]);
+        await waitForAll(['Text:Fallback create passive']);
+        expect(ReactNoop).toMatchRenderedOutput(
+          <>
+            <span hidden={true} prop="One" />
+            <span prop="Fallback" />
+          </>,
+        );
+      });
+
+      // Resolving the suspended resource should re-create inner layout effects.
+      await act(async () => {
+        await resolveText('Two');
+      });
+      assertLog([
+        'AsyncText:Two render',
+        'Text:Fallback destroy insertion',
+        'Text:Fallback destroy layout',
+        'Text:One destroy insertion',
```

Comment:
This one position is what is missing without the fix.

I'm not completely sure how I feel about insertion effects staying connected in hidden subtrees. I think that goes against my intuition as I always thought about them as a layout effect variant. It made sense to me because their introduction is directly related to the needs of CSS-in-JS libraries. 

We have used `useInsertionEffect` in Emotion's `<Global/>` (a component responsible for injecting global styles) and it seems that if those are the desired semantics for `useInsertionEffect` then we should reconsider that. We can't fully rely on the destroy callback being called at the right time here:
https://github.com/emotion-js/emotion/blob/1135f8e9d97ea711eb483368313afdfe7b176845/packages/react/src/global.js#L112-L114

I prepared a demo to showcase the problem **with this patch applied to it** (couldn't use codesandbox's per-commit builds since CSB is 431-ing me when I try to save a sandbox 😢):
https://github.com/Andarist/use-insertion-effect-stays-connected-when-hidden-demo/tree/e11c278fd69c947764d4d5c14399326bff6cdca2

What happens here:
1. Page1 is being loaded (white background since we render fallback)
2. Page1 loads and the background becomes red through the usage of Emotion's `<Global/>`
3. we click the link and navigate to Page2, since it's a lazy component we display the fallback **but** the background stays red while the fallback is being rendered
4. background finally switches to white when Page2 loads

I understand that this might not be how you have wanted us to use this but I'm not sure if this effect here is desirable for us and our users. If insertion effects are meant to be connected in hidden trees then I'll likely have to consider refactoring the `<Global/>` component.